driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java

```diff
@@ -182,249 +218,255 @@ public Boolean getRetryWrites() {
-                            writeRequests, binding.getSessionContext());
-                    return executeBulkWriteBatch(binding, connection, bulkWriteBatch);
+        /* We cannot use the tracking of attempts built in the `RetryState` class because conceptually we have to maintain multiple attempt
+         * counters while executing a single bulk write operation:
+         * - a counter that limits attempts to select server and checkout a connection before we created a batch;
+         * - a counter per each batch that limits attempts to execute the specific batch.
+         * Fortunately, these counters do not exist concurrently with each other. While maintaining the counters manually,
+         * we must adhere to the contract of `RetryingSyncSupplier`. When CSOT is implemented, there will be no counters, and the code
+         * related the attempt tracking in `BulkWriteTracker` will be removed. */
+        RetryState retryState = new RetryState();
+        BulkWriteTracker.attachNew(retryState, retryWrites);
+        Supplier<BulkWriteResult> retryingBulkWrite = decorateWriteWithRetries(retryState, () -> {
+            logRetryExecute(retryState, () -> null);
+            return withSourceAndConnection(binding::getWriteConnectionSource, true, (source, connection) -> {
+                ConnectionDescription connectionDescription = connection.getDescription();
+                int maxWireVersion = connectionDescription.getMaxWireVersion();
+                // attach `maxWireVersion` ASAP because it is used to check whether we can retry
+                retryState.attach(AttachmentKeys.maxWireVersion(), maxWireVersion, true);
+                BulkWriteTracker bulkWriteTracker = retryState.attachment(AttachmentKeys.bulkWriteTracker())
+                        .orElseThrow(Assertions::fail);
+                SessionContext sessionContext = binding.getSessionContext();
+                WriteConcern writeConcern = getAppliedWriteConcern(sessionContext);
+                if (!retryState.firstAttempt() && !isRetryableWrite(retryWrites, writeConcern, source.getServerDescription(),
+                        connectionDescription, sessionContext)) {
+                    RuntimeException prospectiveFailedResult = (RuntimeException) retryState.exception().orElse(null);
+                    retryState.breakAndThrowIfRetryAnd(() -> !(prospectiveFailedResult instanceof MongoWriteConcernWithResponseException));
+                    bulkWriteTracker.batch().ifPresent(bulkWriteBatch -> {
+                        assertTrue(prospectiveFailedResult instanceof MongoWriteConcernWithResponseException);
+                        bulkWriteBatch.addResult((BsonDocument) ((MongoWriteConcernWithResponseException) prospectiveFailedResult)
+                                .getResponse());
+                        BulkWriteTracker.attachNext(retryState, bulkWriteBatch);
+                    });
```

Comment:
This piece of code seems to be dead from the standpoint of the tests we have, i.e., the `breakAndThrowIfRetryAnd` always breaks retrying in tests when it is executed. The same is true for `executeSync`.

The rest of the code seem to be covered.